### PR TITLE
77585 - Create Empty AIO Exchange Set

### PR DESCRIPTION
**Pull request by shailen14599**
Added code changes to create either empty aio/enc or both empty exchange set wrt productVersions endpoint
Added unit tests
Added functional tests
Include additional code changes to fix caching mechanism issue observation ( FileShareServiceCache.cs )

### CODE REVIEW

**Review by ray-sj — CHANGES_REQUESTED** [review]

**Review by ray-sj — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForAioAndProperZipForV01X01WhenAioAndEncProductVersionAndAioIsEnabled.cs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            AuthTokenProvider authTokenProvider = new AuthTokenProvider();
+            EssJwtToken = await authTokenProvider.GetEssToken();
+            FssJwtToken = await authTokenProvider.GetFssToken();
+            ScsApiClient = new SalesCatalogueApiClient(Config.ScsAuthConfig.BaseUrl);
+            ScsJwtToken = await authTokenProvider.GetScsToken();
+            FssApiClient = new FssApiClient();
+            DataHelper = new DataHelper();
+            ProductVersionData = new List<ProductVersionModel>();
+            ProductVersionData.Add(DataHelper.GetProductVersionModelData("DE416080", 9, 1));
+            ProductVersionData.Add(DataHelper.GetProductVersionModelData("GB800001", 31, 34));
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForAioAndProperZipForV01X01WhenAioAndEncProductVersionAndAioIsEnabled.cs]: Changes are done. Please have a look.

Regards,
Mayuresh Gawde

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForAioProductVersionWhenAioIsEnabled.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            AuthTokenProvider authTokenProvider = new AuthTokenProvider();
+            EssJwtToken = await authTokenProvider.GetEssToken();
+            FssJwtToken = await authTokenProvider.GetFssToken();
+            ScsApiClient = new SalesCatalogueApiClient(Config.ScsAuthConfig.BaseUrl);
+            ScsJwtToken = await authTokenProvider.GetScsToken();
+            FssApiClient = new FssApiClient();
+            DataHelper = new DataHelper();
+            ProductVersionData = new List<ProductVersionModel>();
+            ProductVersionData.Add(DataHelper.GetProductVersionModelData("GB800001", 31, 34));
+            ApiEssResponse = await ExchangeSetApiClient.GetProductVersionsAsync(ProductVersionData, accessToken: EssJwtToken);
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForAioProductVersionWhenAioIsEnabled.cs]: Changes are done. Please have a look.

Regards,
Mayuresh Gawde

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForV01X01ProductVersionWhenAioIsEnabled.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+using System.Net.Http;
+
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    class ExchangeSetGeneratesEmptyZipForV01X01ProductVersionWhenAioIsEnabled : ObjectStorage
+    {
+
+        [OneTimeSetUp]
+        public async Task SetupAsync()
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipForV01X01ProductVersionWhenAioIsEnabled.cs:15]: Changes are done. Please have a look.

Regards,
Mayuresh Gawde

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipsForEncAndAioProductVersionWhenAioIsEnabled.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    class ExchangeSetGeneratesEmptyZipsForEncAndAioProductVersionWhenAioIsEnabled : ObjectStorage
+    {
+
+        //Product Backlog Item 77585: ESS : Empty AIO Exchange Set Creation
+        [OneTimeSetUp]
+        public async Task SetupAsync()
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesEmptyZipsForEncAndAioProductVersionWhenAioIsEnabled.cs:19]: Changes are done. Please have a look.

Regards,
Mayuresh Gawde

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesProperZipForAioAndEmptyZipForV01X01WhenAioAndEncProductVersionAndAioIsEnabled.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    class ExchangeSetGeneratesProperZipForAioAndEmptyZipForV01X01WhenAioAndEncProductVersionAndAioIsEnabled : ObjectStorage
+    {
+        
+        //Product Backlog Item 77585: ESS : Empty AIO Exchange Set Creation
+        [OneTimeSetUp]
+        public async Task SetupAsync()
+        {
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGeneratesProperZipForAioAndEmptyZipForV01X01WhenAioAndEncProductVersionAndAioIsEnabled.cs:18]: Changes are done. Please have a look.

Regards,
Mayuresh Gawde

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 6]
```diff
@@ -121,7 +124,17 @@ private Task<BatchDetail> CheckIfCacheProductsExistsInBlob(string exchangeSetRoo
             {
                 var bundleLocation = item.Bundle.FirstOrDefault().Location.Split(";");
                 exchangeSetRootPath = string.Format(exchangeSetRootPath, bundleLocation[0].Substring(1, 1), bundleLocation[1]);
-                downloadPath = Path.Combine(exchangeSetRootPath, item.ProductName.Substring(0, StringLength), item.ProductName, item.EditionNumber.Value.ToString());
+                if (aioConfiguration.IsAioEnabled)
+                {
+                    List<string> aioCells = !string.IsNullOrEmpty(aioConfiguration.AioCells) ? new(aioConfiguration.AioCells.Split(',')) : new List<string>();
+
+                    if (!aioCells.Contains(item.ProductName))
```
Comment (DanOwenUKHO) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:131]: Missing braces

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 7]
```diff
@@ -121,7 +124,17 @@ private Task<BatchDetail> CheckIfCacheProductsExistsInBlob(string exchangeSetRoo
             {
                 var bundleLocation = item.Bundle.FirstOrDefault().Location.Split(";");
                 exchangeSetRootPath = string.Format(exchangeSetRootPath, bundleLocation[0].Substring(1, 1), bundleLocation[1]);
-                downloadPath = Path.Combine(exchangeSetRootPath, item.ProductName.Substring(0, StringLength), item.ProductName, item.EditionNumber.Value.ToString());
+                if (aioConfiguration.IsAioEnabled)
```
Comment (DanOwenUKHO) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:127]: Please split out logic into a seperate method, as this method is becoming very large

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 8]
```diff
@@ -121,7 +124,17 @@ private Task<BatchDetail> CheckIfCacheProductsExistsInBlob(string exchangeSetRoo
...
+                {
+                    List<string> aioCells = !string.IsNullOrEmpty(aioConfiguration.AioCells) ? new(aioConfiguration.AioCells.Split(',')) : new List<string>();
```
Comment (DanOwenUKHO) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:129]: Do we need to instantiate a list here if !string.IsNullOrEmpty(aioConfiguration.AioCells) is false?

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
A few minor points

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
A few minor points

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs (hunk 6 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs]: Resolved. Added braces. Not added braces previously as one liner code was there.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs (hunk 8 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs]: Getting exception if respective property not available in appsettings so added this instantiation.

**Review by sumod13943 — APPROVED** [review]

**Review by ray-sj — APPROVED** [review]

**Review by DanOwenUKHO — APPROVED** [review]

**Pull request merged**